File: 정상영/[240825] 두 큐 합 같게 만들기.js
/*
투 포인터 사용

테스트 1 〉	통과 (0.08ms, 33.4MB)
테스트 2 〉	통과 (0.08ms, 33.4MB)
테스트 3 〉	통과 (0.08ms, 33.4MB)
테스트 4 〉	통과 (0.08ms, 33.4MB)
테스트 5 〉	통과 (0.18ms, 33.4MB)
테스트 6 〉	통과 (0.19ms, 33.4MB)
테스트 7 〉	통과 (0.19ms, 33.4MB)
테스트 8 〉	통과 (0.23ms, 33.4MB)
테스트 9 〉	통과 (0.32ms, 33.4MB)
테스트 10 〉	통과 (0.44ms, 33.6MB)
테스트 11 〉	통과 (12.15ms, 39.7MB)
테스트 12 〉	통과 (9.38ms, 39.4MB)
테스트 13 〉	통과 (8.67ms, 42MB)
테스트 14 〉	통과 (8.93ms, 42.9MB)
테스트 15 〉	통과 (11.75ms, 44.9MB)
테스트 16 〉	통과 (10.18ms, 45.2MB)
테스트 17 〉	통과 (11.61ms, 43MB)
테스트 18 〉	통과 (27.62ms, 59.5MB)
테스트 19 〉	통과 (25.60ms, 60.1MB)
테스트 20 〉	통과 (30.79ms, 60.5MB)
테스트 21 〉	통과 (28.76ms, 60.5MB)
테스트 22 〉	통과 (30.54ms, 60.6MB)
테스트 23 〉	통과 (32.30ms, 60.7MB)
테스트 24 〉	통과 (32.28ms, 60.6MB)
테스트 25 〉	통과 (0.19ms, 33.4MB)
테스트 26 〉	통과 (0.19ms, 33.5MB)
테스트 27 〉	통과 (0.18ms, 33.4MB)
테스트 28 〉	통과 (16.22ms, 43.3MB)
테스트 29 〉	통과 (1.40ms, 34.9MB)
테스트 30 〉	통과 (14.33ms, 44.3MB)
*/

function solution(queue1, queue2) {
    let sum1 = queue1.reduce((acc, cur) => acc + cur, 0);
    let sum2 = queue2.reduce((acc, cur) => acc + cur, 0);
    
    const totalSum = sum1 + sum2;
    if (totalSum % 2 !== 0) return -1;  // 합이 홀수인 경우 불가능
    
    const target = totalSum / 2;
    const n = queue1.length + queue2.length;
    
    let left = 0;  // queue1의 시작 인덱스
    let right = queue1.length; // queue2의 시작 인덱스
    let cnt = 0;

    const queue = [...queue1, ...queue2];

    while (true) {
        if (sum1 === target) return cnt;

        if (sum1 > target) {
            sum1 -= queue[left];
            left++;
        } else {
            sum1 += queue[right];
            right++;
        }

        cnt++;

        // 최대 이동 횟수는 두 큐의 합의 길이 이상을 넘어갈 수 없다
        if (cnt > 2 * n) return -1;
    }

    return -1;
}


/*
시간초과 실패

테스트 1 〉	통과 (0.08ms, 33.5MB)
테스트 2 〉	통과 (0.07ms, 33.4MB)
테스트 3 〉	통과 (0.15ms, 33.6MB)
테스트 4 〉	통과 (0.15ms, 33.6MB)
테스트 5 〉	통과 (0.16ms, 33.5MB)
테스트 6 〉	통과 (0.18ms, 33.2MB)
테스트 7 〉	통과 (0.18ms, 33.2MB)
테스트 8 〉	통과 (0.19ms, 33.5MB)
테스트 9 〉	통과 (0.25ms, 33.5MB)
테스트 10 〉	통과 (0.33ms, 33.7MB)
테스트 11 〉	통과 (1326.60ms, 38.1MB)
테스트 12 〉	통과 (156.16ms, 37.6MB)
테스트 13 〉	통과 (7.31ms, 39.7MB)
테스트 14 〉	통과 (7.38ms, 40.3MB)
테스트 15 〉	통과 (230.22ms, 41.6MB)
테스트 16 〉	통과 (34.02ms, 41.8MB)
테스트 17 〉	통과 (10.73ms, 39.7MB)
테스트 18 〉	통과 (92.50ms, 54.1MB)
테스트 19 〉	통과 (892.56ms, 55.1MB)
테스트 20 〉	통과 (7175.19ms, 56.1MB)
테스트 21 〉	통과 (3093.84ms, 56.4MB)
테스트 22 〉	실패 (시간 초과)
테스트 23 〉	실패 (시간 초과)
테스트 24 〉	실패 (시간 초과)
테스트 25 〉	통과 (0.19ms, 33.4MB)
테스트 26 〉	통과 (0.17ms, 33.6MB)
테스트 27 〉	통과 (0.18ms, 33.4MB)
테스트 28 〉	실패 (시간 초과)
테스트 29 〉	통과 (0.98ms, 36MB)
테스트 30 〉	통과 (5770.59ms, 43.6MB)

function solution(queue1, queue2) {
    let sum1 = 0;
    let sum2 = 0;
    
    queue1.forEach((el) => {
        sum1 += el;
    })
    
    queue2.forEach((el) => {
        sum2 += el;
    })
    
    if((sum1 + sum2) % 2 !== 0) return -1
    
    const middle = (sum1 + sum2) / 2;
    
    let cnt = 0;
    
    while(sum1 !== sum2) {
        if(cnt >= (queue1.length + queue2.length) * 2) return -1;
        
        if(sum1 > sum2) {
            const tmp = queue1.shift();
            sum1 -= tmp;
            queue2.push(tmp);
            sum2 += tmp;
        } else {
            const tmp = queue2.shift();
            sum2 -= tmp;
            queue1.push(tmp);
            sum1 += tmp;
        }
        
        cnt++;
    }
    
    return cnt;
}
*/

File: 이동엽/08월/[240825] 두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();

        long sum1 = 0;
        for (int q : queue1) {
            q1.add(q);
            sum1 += q;
        }

        long sum2 = 0;
        for (int q : queue2) {
            q2.add(q);
            sum2 += q;
        }

        if (sum1 + sum2 % 2 == 1) {
            return -1; // 홀수이면 못나눔
        }

        int cnt = 0;
        int limit = 3 * queue1.length; // 2로하니까 1번 테케 터져서 3으로 하니까 됐음(?)

        while (cnt <= limit) {
            if (sum1 > sum2) {
                int temp = q1.poll();
                q2.add(temp);
                sum1 -= temp;
                sum2 += temp;
            } else if (sum1 < sum2) {
                int temp = q2.poll();
                q1.add(temp);
                sum1 += temp;
                sum2 -= temp;
            } else {
                return cnt;
            }
            cnt++;
        }
        return -1;
    } // method
} // class

//테스트 1 〉	통과 (0.12ms, 73.6MB)
//테스트 2 〉	통과 (0.10ms, 75.5MB)
//테스트 3 〉	통과 (0.11ms, 73.6MB)
//테스트 4 〉	통과 (0.13ms, 73.5MB)
//테스트 5 〉	통과 (0.26ms, 71.1MB)
//테스트 6 〉	통과 (0.42ms, 78.7MB)
//테스트 7 〉	통과 (0.28ms, 71.6MB)
//테스트 8 〉	통과 (0.67ms, 69.8MB)
//테스트 9 〉	통과 (0.87ms, 77.9MB)
//테스트 10 〉	통과 (0.87ms, 72.4MB)
//테스트 11 〉	통과 (17.69ms, 105MB)
//테스트 12 〉	통과 (9.48ms, 82.2MB)
//테스트 13 〉	통과 (11.88ms, 82MB)
//테스트 14 〉	통과 (9.68ms, 92.4MB)
//테스트 15 〉	통과 (11.72ms, 95.4MB)
//테스트 16 〉	통과 (14.80ms, 106MB)
//테스트 17 〉	통과 (11.79ms, 105MB)
//테스트 18 〉	통과 (72.83ms, 160MB)
//테스트 19 〉	통과 (75.98ms, 142MB)
//테스트 20 〉	통과 (75.31ms, 138MB)
//테스트 21 〉	통과 (74.18ms, 148MB)
//테스트 22 〉	통과 (83.23ms, 150MB)
//테스트 23 〉	통과 (80.49ms, 149MB)
//테스트 24 〉	통과 (92.49ms, 151MB)
//테스트 25 〉	통과 (0.49ms, 82.9MB)
//테스트 26 〉	통과 (0.21ms, 77.8MB)
//테스트 27 〉	통과 (0.22ms, 76.2MB)
//테스트 28 〉	통과 (51.36ms, 109MB)
//테스트 29 〉	통과 (2.18ms, 79.1MB)
//테스트 30 〉	통과 (45.94ms, 121MB)